File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args,
        out double val1,
        out CalculatorOperation operation,
        out double val2)
    {
        if (!IsArgLengthSupported(args))
            throw new ArgumentException("The number of arguments is not 3");

        if (!double.TryParse(args[0], out val1))
            throw new ArgumentException("Argument 1 is not a number");
        if (!double.TryParse(args[2], out val2))
            throw new ArgumentException("Argument 2 is not a number");
        operation = ParseOperation(args[1]);
        if (operation is CalculatorOperation.Undefined)
            throw new InvalidOperationException("Unsupported operation");
    }

    private static bool IsArgLengthSupported(IReadOnlyCollection<string> args) => args.Count == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => CalculatorOperation.Undefined
        };
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> CalculatorOperation.Undefined
    
let parseCalcArguments(args : string[]) =
    if not (isArgLengthSupported args) then
        ArgumentException "The number of arguments is not 3"
        |> raise
    let ok, val1 = System.Double.TryParse args[0]
    if not ok then
        ArgumentException "Argument 1 is not a number"
        |> raise
    let ok, val2 = System.Double.TryParse args[2]
    if not ok then
        ArgumentException "Argument 2 is not a number"
        |> raise
    let op = parseOperation args[1]
    if op = CalculatorOperation.Undefined then
        ArgumentException "Unsupported operation"
        |> raise
    {arg1 = val1; arg2 = val2; operation = op}